src/lxc/confile_oci.c

```diff
@@ -291,27 +318,208 @@ static int lxc_oci_linux_idmaps(json_t *elem, struct lxc_conf *conf, char type)
+	}
+
+	return 0;
+}
+
+static int lxc_oci_linux_namespaces(json_t *elem, struct lxc_conf *conf)
+{
+	size_t i;
+	json_t *it;
+
+	if (!json_is_array(elem))
+		return -EINVAL;
+
+	json_array_foreach(elem, i, it) {
+		int ret;
+		const char *type;
+		json_t *val;
+		if (!json_is_object(it))
+			return -EINVAL;
+
+		val = json_object_get(it, "type");
+		if (!json_is_string(val))
+			return -EINVAL;
+
+		type = json_string_value(val);
+		if (strcmp(type, "mount") == 0)
+			type = "mnt";
+		else if (strcmp(type, "network") == 0)
+			type = "net";
+		ret = set_config_namespace_clone("lxc.namespace.clone", type, conf, NULL);
+		if (ret < 0)
+			return ret;
+
+		val = json_object_get(it, "path");
+		if (!val)
+			continue;
+		if (!json_is_string(val))
+			return -EINVAL;
+		WARN("The namespaces \"path\" property is not implemented");
+	}
+
 	return 0;
 }
 
-static int lxc_oci_linux(json_t *elem, struct lxc_conf *conf)
+// https://github.com/opencontainers/runtime-spec/blob/v1.0.1/config-linux.md#default-devices
+// FIXME: use an absolute path for the destination (https://github.com/lxc/lxc/issues/2276)
+static const char * const default_devices[] = {
+	"/dev/null dev/null none bind,nosuid,noexec,create=file 0 0",
+	"/dev/zero dev/zero none bind,nosuid,noexec,create=file 0 0",
+	"/dev/full dev/full none bind,nosuid,noexec,create=file 0 0",
+	"/dev/random dev/random none bind,nosuid,noexec,create=file 0 0",
+	"/dev/urandom dev/urandom none bind,nosuid,noexec,create=file 0 0",
+	"/dev/tty dev/tty none bind,nosuid,noexec,create=file 0 0",
+};
+
+static int lxc_oci_linux_default_devices(struct lxc_conf *conf) {
+	int i;
+	int ret = -1;
+
+	ret = set_config_autodev("lxc.autodev", "0", conf, NULL);
+	if (ret < 0)
+		return ret;
+
+	// For /dev/ptmx
+	ret = set_config_pty_max("lxc.pty.max", "1", conf, NULL);
+	if (ret < 0)
+		return ret;
+
+	for (i = 0; i < sizeof(default_devices) / sizeof(default_devices[0]); ++i) {
+		ret = set_config_mount("lxc.mount.entry", default_devices[i], conf, NULL);
+		if (ret < 0)
+			return ret;
+	}
+
+	return 0;
+}
+
+static int lxc_oci_linux_masked_paths(json_t *elem, struct lxc_conf *conf)
 {
+	size_t i;
+	json_t *it;
+
+	if (!json_is_array(elem))
+		return -EINVAL;
+
+	json_array_foreach(elem, i, it) {
+		int ret;
+		char *entry = NULL;
+		const char *path = NULL;
+
+		if (!json_is_string(it))
+			return -EINVAL;
+
+		path = json_string_value(it);
+		if (path[0] != '/' || path[1] == '\0')
+			return -EINVAL;
+
+		// There is no nice way to do this with LXC today.
+		// 1. Add an optional bind mount of /dev/null, in case the
+		//    target is a file. This will fail if the target is a
+		//    directory.
+		// 2. Add an optional read-only tmpfs mount, in case the target
+		//    is a directory. This will fail if the target is a file.
+		// Unfortunately, since both mounts are optional, we can't
+		// guarantee that one of these mounts will succeed.
+		ret = asprintf(&entry, "/dev/null %s none bind,nosuid,optional 0 0", path + 1);
```

Comment:
This is actually not true. There's a new option `relative` in 3.0 that will allow you to do exactly that. :) See https://github.com/lxc/lxc/pull/2031 and https://github.com/lxc/lxc/blob/a55e2ad107f084ba3b56fde234334f82ad184455/doc/lxc.container.conf.sgml.in#L1078 :)

Reply:
Why make it `optional` then?

Reply:
Not sure what you are suggesting with `relative`, it doesn't matter if I bind-mount the device from the namespace of the runtime or the device from the container rootfs, right?

Reply:
> Why make it optional then?

I don't think I have the choice here. We don't have access to the rootfs at this point so we don't know what to do.
For a file, we need to bind-mount `/dev/null` on top. But doing that on a directory will fail.
For a directory, we need to mount a read-only tmpfs on top. But doing that on a file will fail.

Since both can fail, they need to be both `optional`, in the hope that at least one can succeed. Don't you agree?

Reply:
Hm, ok. We can revisit this later.

Reply:
We can solve that now, if you have a better suggestion :)

Reply:
Let's just merge this in and then work throught the FIXME's together. :)

Reply:
Totally agree! But if they are changes required in the core of LXC, it might take a while to discuss so we should start early.
I will do a little bit more research and open an issue on the main repo.